Fix a few bugs

ob opened the pull request:
* Fix a bug where if record subdirectories already existed, it would
  print an error. fs::create_subdirectories() correctly creates all
  intermediate subdirectories and saves the need to check for
  existence.

* Fix a problem where if the regex passed in the command line didn't
  compile, it would crash. Now it properly reports an error.

kastiglione commented on index-import.cpp:
```diff
@@ -242,7 +225,30 @@ static bool cloneRecords(StringRef recordsDirectory,
+  // Parse the the path remapping command line flags. This converts strings of
+  // "X=Y" into a (regex, string) pair. Another way of looking at it: each
+  // remap is equivalent to the s/pattern/replacement/ operator.
+  auto errors = 0;
```
This can be a `bool` instead of a counter.

ob replied:
Yeah, I had more diffs and in my reshuffle due to clang import dropped some... added now.

kastiglione commented on index-import.cpp:
```diff
@@ -293,6 +299,9 @@ int main(int argc, char **argv) {
       continue;
     }
+    if (Verbose) {
+      outs() << "Remapping file " << unitPath << "\n";
```
👍

kastiglione commented on index-import.cpp:
```diff
@@ -242,7 +225,30 @@ static bool cloneRecords(StringRef recordsDirectory,
+      errors++;
+    }
+    auto replacement = remap.substr(divider + 1);
+    remapper.addRemap(re, replacement);
```
Move these two lines into the `try` block? The `catch` doesn't `continue`, and so these lines are hit even when there's an error. The `try` block could be self contained:

```cpp
try {
  std::regex re{pattern};
  auto replacement = remap.substr(divider + 1);
  remapper.addRemap(re, replacement);
} catch...
```

kastiglione replied:
PS. Nice subtle improvement of `pattern`/`re` naming.

ob replied:
Done

kastiglione commented on index-import.cpp:
```diff
@@ -242,7 +225,30 @@ static bool cloneRecords(StringRef recordsDirectory,
+  }
+
+  if (errors) {
+    errs() << "Aborting due to errors.\n";
```
Throughout the file these are `llvm::errs` and `llvm::outs`. I'm fine with either, but if we drop the `llvm::` scoping, would you mind updating the others to match?

ob replied:
Since we're importing the `llvm` namespace, I went with the plain forms.

kastiglione commented:
> Fix a bug where if record subdirectories already existed, it would print an error.

What was the error?

kastiglione commented on index-import.cpp:
```diff
@@ -242,13 +225,38 @@ static bool cloneRecords(StringRef recordsDirectory,
+    auto pattern = remap.substr(0, divider);
+    std::regex re;
+    try {
+      re = std::regex(pattern);
```
The declaration and initialization can be combined here.

ob replied:
done

ob commented:
> Fix a bug where if record subdirectories already existed, it would print an error.
> What was the error?

I dug a bit more into this and the real bug was not really fixed. 

The bug is that `path::replace_path_prefix()` has a corner case where if you give it two prefixes of the same length, it'll just do string replacement, e.g.  in [this line](https://github.com/lyft/index-import/blob/master/index-import.cpp#L209), if the values for `IndexInputPath` and `OutputIndexPath` are `foo/` and `foo2` respectively, and `outputPath` is `foo/bar/baz`, the resulting `outputPath` will be `foo2bar/baz`.  To reproduce is to run:

```
index-import --remap foo=bar some/path/to/Index/ some/path/to/Index2
```

Thus my replacement of `createdirectory()` for `createdirectories()` only masked out the bug further. I reverted this change and now normalize the paths given in the command line before continuing the processing.

I also split the error handling to print more information about directory creation failure vs file copying failure; while also saving a `stat()` system call by just checking the error returned from the `mkdir` or `copy` operations.

kastiglione commented on index-import.cpp:
```diff
@@ -210,45 +195,89 @@ static bool cloneRecords(StringRef recordsDirectory,
       // doesn't need to be cloned or copied.
-      if (not fs::exists(outputPath)) {
-        failed = fs::copy_file(inputPath, outputPath);
+      failed = fs::copy_file(inputPath, outputPath);
```
Perhaps the comment needs rewording. The `fs::exists` check avoids performing a redundant copy operation, as a performance optimization. Thanks to the content based hash, if `outputPath` exists, then this record file can be skipped because it already exists and has the same contents.

ob replied:
No, I thought `copy_file()` failed when the destination file already existed. I tested and this is not the case... but see my comments on the other RB... we might want both.

kastiglione replied:
I think `fs::copy_file` is part of the problem, the use cases it supports are more narrow. In #15, `fs::copy_file` is no longer used, but it's intended to be temporary. Maybe the right way forward is to use `copyfile` directly. If ever building for linux, then we'd have to add portable fallbacks where cloning doesn't exist.

kastiglione replied:
I'd prefer the "don't copy if exists" logic be explicit, rather implicit behind a `copyfile` flag. In my testing of incremental use (where the output directory already exists), then the two approaches have the same perf.

ob replied:
reverted to explicit logic

kastiglione commented on index-import.cpp:
```diff
@@ -210,45 +195,89 @@ static bool cloneRecords(StringRef recordsDirectory,
 }
 
+// Normalize a path by removing /./ or // from it.
+std::string normalizePath(StringRef Path) {
```
Can this not use `fs::real_path`?

ob replied:
`fs::real_path`` also makes the paths absolute and I _think_ resolves symlinks which is expensive...

kastiglione replied:
Good point on the update. Since it only happens twice, I'd lean toward using `real_path` for simplicity.

ob replied:
Can't use `real_path` on non-existent directories (e.g. the output dir).. I say we stay with the custom code.

kastiglione replied:
Sounds good. Sorry for sending you down a dead end.

ob replied:
np

kastiglione commented:
> The bug is that `path::replace_path_prefix()` has a corner case where if you give it two prefixes of the same length, it'll just do string replacement

thanks for finding and fixing this!

kastiglione commented on index-import.cpp:
```diff
@@ -210,45 +195,89 @@ static bool cloneRecords(StringRef recordsDirectory,
+// Normalize a path by removing /./ or // from it.
+std::string normalizePath(StringRef Path) {
+  SmallString<128> NormalizedPath;
+  for (path::const_iterator I = path::begin(Path), E = path::end(Path); I != E;
```
Can the type be `auto`?

ob replied:
yeah, done.

kastiglione commented:
Thanks!
